nx/lib/nx/binary_backend.ex

```diff
@@ -1038,6 +1038,138 @@ defmodule Nx.BinaryBackend do
     from_binary(out, output_data)
   end
 
+  @impl true
+  def qr(%{shape: {_m, k} = shape_q} = q, %{shape: {k, n}} = r, opts) do
+    mode = opts[:mode]
+
+    keep_reflectors = mode == :raw
+    calculate_q = mode in [:reduced, :complete, :raw]
+
+    q =
+      if calculate_q do
+        identity(q, shape: shape_q)
+      end
+
+    {q, r, reflectors} =
+      for i <- 0..(n - 1), reduce: {q, r, []} do
+        {q, r, reflectors} ->
+          %{shape: {reflector_n, reflector_m}} =
+            reflector = householder_reflector(r[[i..(k - 1), i]])
+
+          shape_h = {k, k}
+          h = identity(as_type(reflector, r), shape: shape_h)
+
+          h_bin =
+            for col <- 0..(reflector_m - 1), row <- 0..(reflector_n - 1), reduce: to_binary(h) do
+              acc ->
+                set_value_at_index(
+                  acc,
+                  h.type,
+                  shape_h,
+                  [row + i, col + i],
+                  reflector[[row, col]]
+                )
+            end
+
+          h = from_binary(h, h_bin)
+
+          reflectors =
+            if keep_reflectors do
+              [h | reflectors]
+            else
+              reflectors
+            end
+
+          q =
+            if calculate_q do
+              Nx.dot(q, h)
+            end
+
+          r = Nx.dot(h, r)
```

Comment:
From above as well, we try to avoid calls to the `Nx` namespace in the backend if at all possible. You can use `bin_zip_reduce` for this if it's just matrix multiplication. If you keep `q`, `r` and `h` as raw binaries then you can copy the matrix multiplication case from the `dot` implementation in this module!